conn.go

```diff
@@ -325,10 +326,53 @@ func (c *Conn) Subprotocol() string {
 	return c.subprotocol
 }
 
-// Close closes the underlying network connection without sending or waiting
-// for a close message.
-func (c *Conn) Close() error {
-	return c.conn.Close()
+// Close sends close frame and waits for one in response
+// it expects two args. `closeCode int` and `closeMessage string` in order
+// it uses variadic args to maintain backwards compatibility
+func (c *Conn) Close(args ...interface{}) error {
+	closeCode := CloseNoStatusReceived
+	message := ""
+	ok := false
+	if len(args) == 2 {
+		closeCode, ok = args[0].(int)
+		if !ok {
+			closeCode = CloseNoStatusReceived
+		}
+		message, ok = args[1].(string)
+		if !ok {
+			message = ""
+		}
+	}
+	err := c.Shutdown(closeCode, message)
+	if err != nil {
+		return err
+	}
+	c.conn.Close()
+	return nil
+}
+
+// Shutdown sends a close frame and waits for one in response
+func (c *Conn) Shutdown(closeCode int, closeMessage string) error {
+	if !isValidCloseCode(closeCode) {
+		// we do not shutdown connection
+		return errors.New("invalid close code received")
+	}
+	if !utf8.ValidString(closeMessage) {
+		return  errors.New("invalid utf8 payload for shutdown message")
+	}
+	message := FormatCloseMessage(closeCode, closeMessage)
+	err := c.WriteControl(CloseMessage, message, time.Now().Add(writeWait))
+	if err != nil {
+		return err
+	}
+	timeStart := time.Now()
+	c.conn.SetReadDeadline(time.Now().Add(time.Minute))
```

Comment:
Because applications typically have a goroutine reading the connection and the read methods cannot be called concurrently with other read methods, the call to Shutdown() must be coordinated with the application's reading goroutine.  That would be OK if the reading goroutine is the goroutine that initiates shutdown, but that's probably not the typical case.

Reply:
but setting read deadline should break that goroutine and (hopefully) give control to us(?)

Reply:
yes, signaling from `advanceFrame` would simplify this a lot

Reply:
The connection is no longer readable after a timeout.